lib/core/src/Cardano/Wallet.hs

```diff
@@ -985,6 +988,55 @@ signPayment ctx wid argGenChange pwd (CoinSelection ins outs chgs) = db & \DBLay
     db = ctx ^. dbLayer @s @k
     tl = ctx ^. transactionLayer @t @k
 
+-- | Makes a fully-resolved coin selection for the given set of payments.
+selectCoinsExternal
+    :: forall ctx s t k e.
+        ( GenChange s
+        , HasDBLayer s k ctx
+        , HasLogger ctx
+        , HasTransactionLayer t k ctx
+        , IsOwned s k
+        , NFData s
+        , Show s
+        , e ~ ErrValidateSelection t
+        )
+    => ctx
+    -> WalletId
+    -> ArgGenChange s
+    -> NonEmpty TxOut
+    -> ExceptT (ErrSelectCoinsExternal e) IO UnsignedTx
+selectCoinsExternal ctx wid argGenChange payments = do
+    CoinSelection mInputs mPayments mChange <-
+        withExceptT ErrSelectCoinsExternalUnableToMakeSelection $
+            selectCoinsForPayment @ctx @s @t @k @e ctx wid payments
+    mOutputs <- db & \DBLayer{..} ->
+        withExceptT ErrSelectCoinsExternalNoSuchWallet $
+            mapExceptT atomically $ do
+                cp <- withNoSuchWallet wid $ readCheckpoint $ PrimaryKey wid
+                (mOutputs, s') <- assignChangeAddresses
+                    argGenChange mPayments mChange $ getState cp
+                putCheckpoint (PrimaryKey wid) (updateState s' cp)
+                pure mOutputs
+    UnsignedTx
+        <$> ensureNonEmpty mInputs  ErrSelectCoinsExternalUnableToAssignInputs
+        <*> ensureNonEmpty mOutputs ErrSelectCoinsExternalUnableToAssignOutputs
```

Comment:
**[remark]** 

I am not fond of having `xxxUnableToAssignxxx` at the value-level here because, these are clearly invariant violation. If you give a non-empty list of output to the coin selection (as we do) then the result MUST have non empty inputs and outputs. Anything else is an invariant violation and ought to be treated as an exceptional case (i.e. with an exception). 

In a sense, this is a consequence of our "insufficient" typing regarding `CoinSelection`. A possible "easy fix" would be to simply move that contract inside the coin selection functions (`random` and `largestFirst`) adding invariant about the shape of the result.

Reply:
> I am not fond of having `xxxUnableToAssignxxx` at the value-level here because, these are clearly invariant violation. If you give a non-empty list of output to the coin selection (as we do) then the result MUST have non empty inputs and outputs. Anything else is an invariant violation and ought to be treated as an exceptional case (i.e. with an exception).
> 
> In a sense, this is a consequence of our "insufficient" typing regarding `CoinSelection`. A possible "easy fix" would be to simply move that contract inside the coin selection functions (`random` and `largestFirst`) adding invariant about the shape of the result.

I completely agree, and think it would be better to have the "cannot be empty" invariants encoded in the `CoinSelection` type itself.

I'm guessing this would require quite a bit of refactoring. Perhaps we should put this on our list to tidy up during a recovery week?